tests/devices/test_preprocess.py

```diff
@@ -512,6 +517,266 @@ def test_error_incompatible_mcm_method(self):
             _, _ = mid_circuit_measurements(tape, dev, mcm_config)
 
 
+class TestMeasurementsFromCountsOrSamples:
+    """Tests for transforms modifying measurements to be derived from either counts or samples"""
+
+    @pytest.mark.parametrize(
+        "meas_transform", (measurements_from_counts, measurements_from_samples)
+    )
+    def test_error_without_valid_wires(self, meas_transform):
+        """Test that a clear error is raised if the transform fails because a measurement
+        without wires is passed to it"""
+
+        tape = qml.tape.QuantumScript([], measurements=[qml.probs()], shots=10)
+
+        with pytest.raises(
+            RuntimeError,
+            match="Please apply validate_device_wires transform before measurements_from_",
+        ):
+            meas_transform(tape)
+
+    # pylint: disable=unnecessary-lambda
+    @pytest.mark.parametrize(
+        "meas_transform", (measurements_from_counts, measurements_from_samples)
+    )
+    @pytest.mark.parametrize(
+        "input_measurement, expected_res",
+        [
+            (
+                qml.expval(qml.PauliY(wires=0) @ qml.PauliY(wires=1)),
+                lambda theta: np.sin(theta) * np.sin(theta / 2),
+            ),
+            (qml.var(qml.Y(wires=1)), lambda theta: 1 - np.sin(theta / 2) ** 2),
+            (
+                qml.probs(),
+                lambda theta: np.outer(
+                    np.outer(
+                        [np.cos(theta / 2) ** 2, np.sin(theta / 2) ** 2],
+                        [np.cos(theta / 4) ** 2, np.sin(theta / 4) ** 2],
+                    ),
+                    [1, 0, 0, 0],
+                ).flatten(),
+            ),
+            (
+                qml.probs(wires=[1]),
+                lambda theta: [np.cos(theta / 4) ** 2, np.sin(theta / 4) ** 2],
+            ),
+        ],
+    )
+    @pytest.mark.parametrize("shots", [3000, (3000, 3000), (3000, 3500, 4000)])
+    def test_measurements_from_samples_or_counts(
+        self,
+        meas_transform,
+        input_measurement,
+        expected_res,
+        shots,
+    ):
+        """Test the test_measurements_from_samples and measurements_from_counts transforms with a
+        single measurement, and compare outcome to the analytic result.
+        """
+
+        theta = 2.5
+        dev = qml.device("default.qubit", wires=4, shots=shots)
+
+        tape = qml.tape.QuantumScript(
+            [qml.RX(theta, 0), qml.RX(theta / 2, 1)],
+            measurements=[input_measurement],
+            shots=shots,
+        )
+        (validated_tape,), _ = validate_device_wires(tape, dev.wires)
+        tapes, fn = meas_transform(validated_tape)
```

Comment:
```suggestion
        tapes, fn = meas_transform(validated_tape)
        assert len(tapes) == 1
        expected_m = qml.sample(wires=input_measurement.wires) if meas_transform == measurements_from_samples else qml.counts(wires=input_measurement.wires, all_outcomes=False)
        assert len(tapes[0].measurements) == 1
        qml.assert_equal(tapes[0].measurements[0], expected_m)
```
Or something like that.